feat(es): 实现基于TTL的映射缓存机制

Zhaoyikaiii opened the pull request:
- 在Query结构体中添加FieldsUniqueKey方法用于生成缓存键
- 实现MappingCache类型，支持Put、Get、Delete和Clear操作
- 在Instance结构体中添加mappingCache字段
- 设置默认MappingCacheTTL常量为5分钟
- 在mappingCache中添加MappingTTL选项支持自定义TTL配置
- 在QueryRawData和QuerySeriesSet方法中集成缓存机制
- 添加完整的TestInstance_mappingCache测试用例验证缓存功能

EASYGOING45 committed:
fix: 修复共享集群过滤无效问题 --bug=129763639 (#505)

EASYGOING45 committed:
feat: BCS共享集群-添加更多日志 --story=119412655 (#507)

chenjiandongx committed:
sync: master branch (#509)

Co-authored-by: LDepression <1197285120@qq.com>
Co-authored-by: frostycliu <frostycliu@tencent.com>
Co-authored-by: Shamcle Ren <shamcleren@tencent.com>
Co-authored-by: ctenetlau <74183504+EASYGOING45@users.noreply.github.com>

shamcleren committed:
feat: 解决 bksql 鉴权参数为空的问题 --story=119438132 (#511)

shamcleren committed:
Merge branch 'release'

shamcleren committed:
Merge branch 'release'

shamcleren committed:
Merge branch 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branch 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branch 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branch 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branch 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branches 'master' and 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branches 'master' and 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branch 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branches 'master' and 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branches 'master' and 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branch 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branch 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branch 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branches 'master' and 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branch 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branch 'master' of github.com:TencentBlueKing/bkmonitor-datalink

Zhaoyikaiii committed:
feat(es): 实现基于TTL的映射缓存机制

- 在Query结构体中添加FieldsUniqueKey方法用于生成缓存键
- 实现MappingCache类型，支持Put、Get、Delete和Clear操作
- 在Instance结构体中添加mappingCache字段，使用RWMutex保证线程安全
- 设置默认MappingCacheTTL常量为5分钟
- 在InstanceOption中添加MappingTTL选项支持自定义TTL配置
- 在QueryRawData和QuerySeriesSet方法中集成缓存机制
- 添加完整的TestInstance_mappingCache测试用例验证缓存功能

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -53,6 +57,10 @@ type Instance struct {
 	maxSize int
+
+	mappingCache     MappingCache
+	mappingCacheLock sync.RWMutex
```
锁的实现是不是可以下沉到 cache 本身，因为看起来每个 instance 也是一个 map

Zhaoyikaiii replied:
已修改。把这个全局锁下沉到了cache。

Zhaoyikaiii committed:
feat(es): 优化映射缓存的并发安全性

1. 重构 MappingCache 类型，将其从简单的 map 改为包含 map 和锁的结构体
2. 为 MappingCache 添加自己的读写锁，实现更细粒度的并发控制
3. 更新相关方法，移除全局锁，使用缓存内部的锁机制
4. 增加并发测试用例，验证新实现在高并发场景下的正确性和性能

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/mapping_cache.go:
```diff
@@ -27,81 +28,123 @@ func (m MappingEntry) IsExpired(ttl time.Duration) bool {
 
 // Put 添加映射到缓存
 func (m *MappingCache) Put(tableID string, fieldsStr string, mappings []map[string]any) {
 	if m == nil {
```
倒是觉得这里可以不用，这相当于所有 pointer receiver 都需要再调用之前先判断自己是否为空指针，而一般不会这么做，不然代码写起来心智负担就太大了。

chenjiandongx replied:
这里看起来好像没有修改？后续不再赘述，请再都检查一遍。

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/mapping_cache.go:
```diff
@@ -27,81 +28,123 @@ func (m MappingEntry) IsExpired(ttl time.Duration) bool {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.data == nil {
```
另外，这里什么情况下会为空；还有为什么 NewMappingCache 返回的是 struct，而所有签名都是 *struct？

Zhaoyikaiii replied:
这里保留这个检查是一种防御性编程的做法，可以防止两种情况：
1. 没有通过 `NewMappingCache()` 而直接创建了一个零值 `MappingCache` 实例
2. 外部代码可能会手动将 `data` 设置为 `nil`

chenjiandongx replied:
所以你在担心一个 Exported 结构体的小写变量会在外部被人置空？同时你的实现还要时刻保证其自身的 pointer 不为空？当然，没说这样做是错的，我想表达的只是一般编码不会这样写，如果按照这个规则，那这个项目的所有指针引用都需要这么重改一遍了。

拿最基础的 `cli := &http.Client{}` 来说，所有的使用都应该是 *struct，但其内部实现是必不会为每个 Func 都增加 nil 判断的。

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/mapping_cache.go:
```diff
@@ -27,81 +28,123 @@ func (m MappingEntry) IsExpired(ttl time.Duration) bool {
 }
 
 // Get 从缓存获取映射条目，自动处理过期条目
 func (m *MappingCache) Get(tableID string, fieldsStr string, ttl time.Duration) (MappingEntry, bool) {
```
可以考虑把读锁和写锁分成两个函数，这个锁分了 4 5 个地方操作，后续改动也容易出错。

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -129,7 +128,7 @@ func NewInstance(ctx context.Context, opt *InstanceOption) (*Instance, error) {
 		healthCheck: opt.HealthCheck,
 		timeout:     opt.Timeout,
 
-		mappingCache: make(MappingCache),
```
一般如果会改动 struct 本身变量的 Exported 的结构体，一般都建议以指针使用。

Zhaoyikaiii replied:
已修改为指针类型。

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -58,9 +58,8 @@ type Instance struct {
-	mappingCacheLock sync.RWMutex
-	mappingTTL       time.Duration
+	mappingCache MappingCache
+	mappingTTL   time.Duration
```
如果 instance 本身不会在调用的时候传递不同的 ttl，也可以考虑将 ttl 直接在 mapping 初始化的时候直接赋值。

Zhaoyikaiii committed:
feat: 将MappingCache receiver改为使用pointer

1. 将 NewMappingCache() 函数返回值从值类型改为指针类型 (*MappingCache)
2. 更新 Instance 结构体中 mappingCache 字段类型为 *MappingCache

Zhaoyikaiii committed:
优化: 重构MappingCache使用辅助函数封装锁操作

1. 添加了withReadLock和withWriteLock两个helper function，用于封装读写锁的操作
2. 修改所有方法(Put/Get/Delete/Clear)使用这些辅助函数替代直接操作锁

Zhaoyikaiii committed:
feat: 移除instance的ttl到mappingCache下

Zhaoyikaiii committed:
feat: 移除mappingCache的Put操作时的判空检查

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/mapping_cache.go:
```diff
@@ -0,0 +1,250 @@
+	tableID := parts[0]
+	fieldsStr := ""
+	if len(parts) > 1 {
+		fieldsStr = strings.Join(parts[1:], "|")
```
这里的 key 设计也可以优化，你先是计算时进行了拼接，然后读取后进行拆分，判断后又重新拼接。那为什么不一开始就分为 tableid 和 field？

Zhaoyikaiii replied:
已修改。GetCacheKey会分别返回tableId和fieldStr。

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -713,11 +733,20 @@ func (i *Instance) QuerySeriesSet(
+					}
+					if len(mappings) == 0 {
+						log.Warnf(ctx, "index is empty with %v", qo.indexes)
+						err = fmt.Errorf("index is empty with %v，url: %s", aliases, conn.Address)
```
这里需要打印 address 的原因是什么？

Zhaoyikaiii committed:
refactor(struct): 移除旧方法并新增生成缓存键的方法

Zhaoyikaiii committed:
chore: 修改检查获取缓存参数tableID和fieldStr

Zhaoyikaiii committed:
refactor(mapping_cache.go): 移除TTL操作的锁，简化赋值和返回

Zhaoyikaiii committed:
refactor(mapping_cache): 优化映射缓存读取逻辑并新增辅助方法

Zhaoyikaiii committed:
test: 增强并优化 Elasticsearch 映射缓存功能的测试

Zhaoyikaiii committed:
refactor(instance): 移除索引为空时错误信息的 URL 字段

Zhaoyikaiii committed:
chore: 移除MappingCache在withWriteLock时的判空处理

Zhaoyikaiii committed:
feat: mappingCache重用withWriteLock,withReadLock辅助函数

shamcleren commented on pkg/unify-query/metadata/struct.go:
```diff
@@ -379,3 +379,40 @@ func (os Orders) SortSliceList(list []map[string]any) {
+// GetCacheKey 返回tableID和fields，用于缓存查询
+func (q *Query) GetCacheKey() (tableID string, fieldsStr string) {
+	if q.TableID == "" {
+		return "", ""
```
这里可以直接写成 return

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -516,11 +530,24 @@ func (i *Instance) QueryRawData(ctx context.Context, query *metadata.Query, star
+				// 缓存新获取的映射
+				err := i.writeMappingCache(mappings, tableID, fieldsStr)
+				if err != nil {
+					err = mappingErr
```
那这个真实的 err 是不是会被吃掉。

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -516,11 +530,24 @@ func (i *Instance) QueryRawData(ctx context.Context, query *metadata.Query, star
+			tableID, fieldsStr := query.GetCacheKey()
+			mappings, exist := i.checkMappingCache(tableID, fieldsStr)
+			if exist {
+				span.Set("mapping-exist", true)
```
这里可以更简洁一点，直接传入 exist 作为布尔值即可。

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance_test.go:
```diff
@@ -747,3 +747,367 @@ func TestInstance_queryRawData(t *testing.T) {
+		MappingTTL: 100 * time.Millisecond, // 设置较短的TTL方便测试过期
+	})
+	if err != nil {
+		t.Logf("Instance creation error: %v - continuing with test", err)
```
这里即使错误也要求继续往下测试？

chenjiandongx replied:
我的建议是这里直接抛错，终止测试，而不是出错输出日志直接返回。

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance_test.go:
```diff
@@ -747,3 +747,367 @@ func TestInstance_queryRawData(t *testing.T) {
+
+		for _, c := range combinations {
+			err := ins.writeMappingCache(sampleMappings, c.tableID, c.fieldsStr)
+			assert.Nil(t, err, "Failed to write mappings to cache for %s|%s", c.tableID, c.fieldsStr)
```
现在应该就不是 '%s|%s' 的格式了。

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance_test.go:
```diff
@@ -747,3 +747,367 @@ func TestInstance_queryRawData(t *testing.T) {
+			fieldStrs[fieldStr] = true
+		}
+
+		assert.Equal(t, 2, len(tableIDs), "Should have 2 unique tableIDs")
```
可以用 `assert.Len` 方法

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance_test.go:
```diff
@@ -747,3 +747,367 @@ func TestInstance_queryRawData(t *testing.T) {
+		MappingTTL: 100 * time.Millisecond,
+	})
+	if err != nil {
+		t.Logf("Instance creation error: %v - continuing with test", err)
```
同上。

Zhaoyikaiii committed:
refactor(instance): 简化映射关系获取逻辑

Zhaoyikaiii committed:
refactor(instance): 改进 writeMappingCache 错误处理逻辑

Zhaoyikaiii committed:
fix: 实例创建错误时提前返回，避免测试受影响

Zhaoyikaiii committed:
test: 更新测试代码断言方式

Zhaoyikaiii committed:
refactor(mapping_cache.go): 移除 `MappingCache` 方法中对 `m == nil` 的检查

Zhaoyikaiii committed:
test: 改进 instance_test.go 中断言消息格式

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -516,11 +530,23 @@ func (i *Instance) QueryRawData(ctx context.Context, query *metadata.Query, star
+			tableID, fieldsStr := query.GetCacheKey()
+			mappings, exist := i.checkMappingCache(tableID, fieldsStr)
+			span.Set("mapping-exist", exist)
+			if exist {
```
那这个 if 留着干嘛呢？

Zhaoyikaiii committed:
refactor(instance): 优化缓存映射存在性检查逻辑

Zhaoyikaiii committed:
fix: 终止实例创建错误的测试

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -37,6 +37,10 @@ import (
 )
 
+const (
+	MappingCacheTTL = 5 * time.Minute
```
这里的超时时间改成可配置的

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -106,6 +113,11 @@ var TimeSeriesResultPool = sync.Pool{
 func NewInstance(ctx context.Context, opt *InstanceOption) (*Instance, error) {
+	mappingTTL := opt.MappingTTL
+	if mappingTTL <= 0 {
+		mappingTTL = MappingCacheTTL
```
这里如果要这样用的话，应该改名成 DefaultMappingCacheTTL

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -516,11 +530,22 @@ func (i *Instance) QueryRawData(ctx context.Context, query *metadata.Query, star
+				// 缓存新获取的映射
+				err := i.writeMappingCache(mappings, tableID, fieldsStr)
+				if err != nil {
+					err = fmt.Errorf("write mapping cache failed: %v, original error: %v", err, mappingErr)
```
这里最好改成 warn，不影响主流程

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/mapping_cache.go:
```diff
@@ -0,0 +1,193 @@
+		}
+
+		m.data[tableID][fieldsStr] = MappingEntry{
+			mappings:    mappings,
```
这里为什么要把整个 mappings 加到这个字段下？只需要缓存当前 tableID + field 下的类型就行了吧？

Zhaoyikaiii committed:
refactor(instance): Rename update mapping cache TTL

Zhaoyikaiii committed:
refactor(struct.go): Modify GetCacheKey to return zero values on empty TableID

Zhaoyikaiii committed:
refactor: Refactor mapping and field type handling
# Reviewed, transaction id: 41595

Zhaoyikaiii committed:
refactor(instance): Rename mappingCache to fieldTypesCache
# Reviewed, transaction id: 41597

Zhaoyikaiii committed:
feat: Add methods and field for mapping cache
# Reviewed, transaction id: 41602

Zhaoyikaiii committed:
refactor: Improve cache handling and factory mocking
# Reviewed, transaction id: 41606

Zhaoyikaiii committed:
refactor: Refactor field name and test mappings
# Reviewed, transaction id: 41618

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -301,10 +309,13 @@ func (f *FormatFactory) WithOrders(orders metadata.Orders) *FormatFactory {
-func (f *FormatFactory) WithMappings(mappings ...map[string]any) *FormatFactory {
+func (f *FormatFactory) WithMappings(tableID string, fieldTypesFactory FieldTypeFactory, mappings ...map[string]any) *FormatFactory {
 	for _, mapping := range mappings {
 		mapProperties("", mapping, f.mapping)
```
这里的 f.mapping 可以移除掉了，只会用来获取字段类型，所以用 fieldTypesCache 已经可以满足需要了

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/instance.go:
```diff
@@ -114,6 +126,8 @@ func NewInstance(ctx context.Context, opt *InstanceOption) (*Instance, error) {
 		healthCheck: opt.HealthCheck,
 		timeout:     opt.Timeout,
+
+		fieldTypesCache: NewMappingCache(mappingTTL),
```
这里的 MappingCache 应该是维护一个全局的缓存吧？

Zhaoyikaiii committed:
refactor: Refactor struct and instance tests

- remove GetCacheKey function

Zhaoyikaiii committed:
refactor: change fieldTypesCache as a global instance
# Reviewed, transaction id: 41731

Zhaoyikaiii committed:
refactor: Refactor mapping cache functionality tests

- remove instance cache test sample
- add mapping cache test
# Reviewed, transaction id: 41737

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/mapping_cache.go:
```diff
@@ -0,0 +1,189 @@
+)
+
+func init() {
+	fieldTypesCache = *NewMappingCache(DefaultMappingCacheTTL)
```
为什么会有这种 `xx = *NewXxx` 的这种用法？目的是什么呢

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/mapping_cache.go:
```diff
@@ -0,0 +1,189 @@
+// NewMappingCache 创建一个新的映射缓存
+func NewMappingCache(ttl time.Duration) *MappingCache {
+	return &MappingCache{
+		data: make(map[string]map[string]MappingEntry),
```
如果所有的操作都是需要 table + field，那我建议还是使用一个两者的结构体作为 key。

```
struct{ table, field string}
```

这样即不用每次都使用 `table+field` 两重判断；其次就不用生成两重 map

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/mapping_cache_test.go:
```diff
@@ -0,0 +1,386 @@
+	// Wait for all goroutines to complete
+	wg.Wait()
+
+	// If we reach here without deadlocks or panics, the test is successful
```
这种是 AI 生成的？感觉写在这里有点奇怪，因为好像每个测试都应该是这个样子。

Zhaoyikaiii committed:
refactor(mapping_cache): Enhance logging in GetFieldType
# Reviewed, transaction id: 41949

Zhaoyikaiii committed:
refactor(mapping_cache): Refactor structure and operations
# Reviewed, transaction id: 41957

Zhaoyikaiii committed:
refactor: remove commit comments
# Reviewed, transaction id: 41963

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/mapping_cache.go:
```diff
@@ -0,0 +1,166 @@
+)
+
+func init() {
+	fieldTypesCache = NewMappingCache(DefaultMappingCacheTTL)
```
如果希望全局变量是指针，那在 var 声明时写为 `fieldTypesCache *MappingCache` 即可，也不是什么大问题。

Zhaoyikaiii committed:
chore: group import packages

Zhaoyikaiii committed:
chore: add GetFieldType log

Zhaoyikaiii committed:
test: Add multi-query scenario tests

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/mapping_cache.go:
```diff
@@ -0,0 +1,170 @@
+}
+
+func (m *MappingCache) get(tableID string, fieldsStr string) (MappingEntry, bool) {
+	return m.cleanupOrGetEntry(tableID, fieldsStr)
```
建议清理缓存逻辑，单独启一个协程处理，不然如果没有查询触发，这里的缓存就永远不会清理了，或者使用开源的缓存库

Zhaoyikaiii committed:
refactor(mapping_cache): Refactor cache implementation

- using go-cache

Zhaoyikaiii committed:
feat(cache): Replace go-cache with ristretto and upgrade dependencies

shamcleren committed:
Merge branches 'master' and 'master' of github.com:TencentBlueKing/bkmonitor-datalink

chenjiandongx commented on pkg/unify-query/tsdb/elasticsearch/mapping_cache.go:
```diff
@@ -0,0 +1,103 @@
+func (m *MappingCache) GetFieldType(tableID string, fieldsStr string) (string, bool) {
+	key := createCacheKey(tableID, fieldsStr)
+
+	fieldType, found := m.cache.Get(key)
```
直接 return 看起来就可以了。

Zhaoyikaiii committed:
style: Adjust formatting in client.go and reorder import in mapping_cache.go

Zhaoyikaiii committed:
refactor(mapping_cache): Simplify GetFieldType method by removing unnecessary checks

Zhaoyikaiii committed:
refactor(mapping_cache): Update cache configuration to use viper for dynamic settings

Zhaoyikaiii committed:
refactor(mapping_cache): Update methods to include context parameter for better traceability

- Modified GetFieldType, AppendFieldTypesCache, Delete, and Clear methods to accept context.Context.
- Updated related tests to pass context and ensure functionality remains intact.

Zhaoyikaiii committed:
refactor(mapping_cache): Improve tracing in cache methods

- Updated span creation in AppendFieldTypesCache, GetFieldType, Delete, and Clear methods to remove context variable.
- Enhanced cache hit logging in GetFieldType method to differentiate between cache hits and misses.

Zhaoyikaiii committed:
refactor(mapping_cache): Enhance tracing by logging mapping keys in AppendFieldTypesCache

- Added logging of mapping keys in the AppendFieldTypesCache method to improve traceability and debugging.

shamcleren committed:
Merge branch 'master' into feat/es_mapping_cache

Zhaoyikaiii committed:
refactor(mapping_cache): Simplify cache hit logging in GetFieldType method

shamcleren committed:
Merge branch 'master' of github.com:TencentBlueKing/bkmonitor-datalink

shamcleren committed:
Merge branch 'bkop' into feat/es_mapping_cache

Zhaoyikaiii committed:
Feat/es collapse feature/#1010158081123984180 (#769)

Co-authored-by: Shamcle Ren <shamcleren@tencent.com>

Zhaoyikaiii committed:
Merge branch 'bkop' into feat/es_mapping_cache

Zhaoyikaiii committed:
refactor: extract query source setup into a dedicated function for collapse handling

- Introduced `setupCollapseQuerySource` function to streamline the construction of Elasticsearch queries with collapse aggregation.
- Updated test cases to utilize the new function, improving code readability and maintainability.
- Adjusted expected results in tests to reflect correct collapse field usage.

Implements Elasticsearch mapping caching

Implements caching for Elasticsearch mappings to improve query performance.

This change introduces a new mapping cache that stores the complete mapping information for Elasticsearch indices, which can then be retrieved when needed, reducing the load on the Elasticsearch cluster.

Zhaoyikaiii committed:
chore: update indirect dependencies in go.mod

- Upgraded `golang.org/x/sync` to v0.11.0.
- Updated `golang.org/x/text` to v0.22.0.
- Changed `golang.org/x/tools` to v0.21.1-0.20240508182429-e35e4ccd0d2d.
